lib/engine/game/g_1824.rb

```diff
@@ -228,41 +237,126 @@ def init_round
-        coal_railways.each do |coalcorp|
-          train = g_trains.shift
-          buy_train(coalcorp, train, :free)
-          coalcorp.spend(120, @bank, check_cash: false)
-        end
+        @two_train_bought = false
 
         @minors.each do |minor|
           hex = hex_by_id(minor.coordinates)
           hex.tile.cities[minor.city].place_token(minor, minor.next_token)
         end
+
+        # Reserve the last share (which is the 20% presidency) to
+        # have it as exchange for associated coal railway
+        associated_regional_railways.each do |regional|
+          regional.shares.find(&:president).buyable = false
+          regional.floatable = false
+        end
       end
 
       def ipo_name(entity)
-        return 'Treasury' if coal_railways.include?(entity)
+        return 'Treasury' if coal_railway?(entity)
 
         'IPO'
       end
 
       def can_par?(corporation, parrer)
-        super && !corporation.all_abilities.find { |a| a.type == :no_buy }
+        super && !corporation.all_abilities.find { |a| a.type == :no_buy } &&
+          !associated_regional_railways.include?(corporation) &&
+          !(coal_railway?(corporation) && corporation.floated?)
       end
 
       def g_train?(train)
         train.name.end_with?('g')
       end
 
-      def coal_railways
-        @coal_railways ||= (option_cisleithania ? %w[EPP EOD MLB] : %w[EPP EOD MLB SPB])
-          .map { |c| corporation_by_id(c) }
+      def mountain_railway?(entity)
+        entity.company? && entity.sym.start_with?('B')
+      end
+
+      def coal_railway?(entity)
+        entity.corporation? && entity.type == :Coal
+      end
+
+      def pre_staatsbahn?(entity)
+        entity.minor? && entity.type == :PreStaatsbahn
+      end
+
+      def regional?(entity)
+        entity.corporation? && entity.type == :Regional
+      end
+
+      def staatsbahn?(entity)
+        entity.corporation? && entity.type == :Staatsbahn
+      end
+
+      def associated_regional_railways
+        return @associated_regional_railways if @associated_regional_railways
+
+        @associated_regional_railways = [regional_bk, regional_cl, regional_ms]
+        @associated_regional_railways << regional_sb unless option_cisleithania
+        @associated_regional_railways
+      end
+
+      def associated_regional_railway(coal_railway)
+        case coal_railway.name
+        when 'EPP'
+          regional_bk
+        when 'EOD'
+          regional_ms
+        when 'MLB'
+          regional_cl
+        when 'SPB'
+          regional_sb
+        end
+      end
+
+      def associated_state_railway(prestate_railway)
+        case prestate_railway.id
+        when 'S1', 'S2', 'S3'
+          state_sd
+        when 'U1', 'U2'
+          state_ug
+        when 'K1', 'K2'
+          state_kk
+        end
+      end
+
+      def operating_order
+        @corporations.select(&:floated?).select { |c| coal_railway?(c) } +
```

Comment:
combined the selects so you don't do two passes

Reply:
Changed to use partition instead